clang/lib/Tooling/DependencyScanning/IncludeTreeActionController.cpp

```diff
@@ -59,7 +59,7 @@ class IncludeTreeActionController : public CallbackActionController {
   // pointer so the builder cannot move when resizing.
   SmallVector<std::unique_ptr<IncludeTreeBuilder>> BuilderStack;
   std::optional<cas::IncludeTreeRoot> IncludeTreeResult;
-  llvm::StringMap<std::string> OutputToCacheKey;
+  llvm::StringMap<std::optional<std::string>> OutputToCacheKey;
```

Comment:
Why is this optional instead of not storing in the case it would be nullopt?

Reply:
I wanted to distinguish between "have we seen a command with this output path before" and "did we successfully create the cache key for that command". I think that this should not actually matter much - we should make both of these cases impossible.

I might've been more defensive than necessary due to the fact that I preserved the logic that ignores the failure to create a cache key for a command. Original from`DependencyScanningWorker.cpp` below:

```c++
    if (ScanInstance.getFrontendOpts().CacheCompileJob) {
       auto &CAS = ScanInstance.getOrCreateObjectStore();
       if (auto Key = createCompileJobCacheKey(
               CAS, ScanInstance.getDiagnostics(), OriginalInvocation))
         TUCacheKey = Key->toString();
     }
```

Here, `createCompileJobCacheKey()` function would turn any internal errors (`llvm::Expected` and the absence of either `FrontendOptions::CASIncludeTreeID`/`FileSystemOptions::CASFileSystemRootID`) into a `std::nullopt` result (and emit an error diagnostic).

This call is now made in `IncludeTreeActionController::finalize()`. Maybe I should fail early in that case and propagate the error up all the way to `forEachDriverJob()` that guarantees early termination of the loop over all commands. Then I'd be more comfortable squashing the optionality out of `OutputToCacheKey` entries and asserting more aggressively on the presence of expected entries. WDYT?

Reply:
Turns out we create tons of distinct `ObjectStore` instances (mainly from unit-tests) and propagating errors out makes them fail. I'll try to fix that in a follow-up PR.